worldedit-sponge/src/main/java/com/sk89q/worldedit/sponge/SpongeWorld.java

```diff
@@ -142,6 +152,43 @@ public boolean setBlock(Vector position, BaseBlock block, boolean notifyAndLight
 
     @Override
     public boolean regenerate(Region region, EditSession editSession) {
+        try {
+            Server server = Sponge.getServer();
+            WorldProperties tempWorldProperties;
+            try {
+                tempWorldProperties = server.createWorldProperties("worldedittemp", this.regenerationArchetype);
+            } catch (IOException e) {
+                e.printStackTrace();
+                return false;
+            }
+
+            Optional<World> tempWorldOpt = server.loadWorld(tempWorldProperties);
+            if (!tempWorldOpt.isPresent()) {
+                return false;
+            }
+
+            World freshWorld = tempWorldOpt.get();
+
+            // Pre-gen all the chunks
+            // We need to also pull one more chunk in every direction
+            CuboidRegion expandedPreGen = new CuboidRegion(region.getMinimumPoint().subtract(16, 0, 16), region.getMaximumPoint().add(16, 0, 16));
+            for (Vector2D chunk : expandedPreGen.getChunks()) {
+                freshWorld.getChunk(chunk.getBlockX(), 0, chunk.getBlockZ());
+            }
+
+            SpongeWorld from = SpongeWorldEdit.inst().getWorld(freshWorld);
+            for (BlockVector vec : region) {
+                editSession.setBlock(vec, from.getBlock(vec));
+            }
+
+            // Remove temp world
+            server.unloadWorld(freshWorld);
```

Comment:
Might want to put this (and following line) in a `finally` clause, in case the `MaxChangedBlocksException` occurs.